drafts/draft-svaldez-pp-http-api/draft-svaldez-pp-http-api.md

```diff
@@ -0,0 +1,423 @@
+
+Inputs:
+- ``server_origin``: The origin to retrieve a key commitment for.
+
+No outputs.
+
+1. The client fetches the configuration state ``server_id``,
+   ``ciphersuite``, ``commitment_id`` associated with ``server_origin``.
+
+2. The client makes an anonymous GET request to
+   ``CONFIG_ENDPOINT``/.well-known/privacy-pass with a message of type
+   ``fetch-commitment`` and a body of:
+
+~~~
+struct {
+    opaque server_id<1..2^16-1> = server_id;
+    opaque commitment_id<1..2^8-1> = commitment_id;
+}
+~~~
+
+3. The server looks up the current configuration, and constructs a list
+of commitments to return, noting whether a key commitment is valid for
+issuance or redemption or both.
+
+4. The server then responds with a message of type ``commitment`` and a
+   body of:
+
+~~~
+struct {
+    opaque public_key<1..2^16-1>;
+    uint64 expiry;
+    uint8 supported_methods; # 3 = Issue/Redeem, 2 = Redeem, 1 = Issue
+    opaque signature<1..2^16-1>;
+} KeyCommitment;
+
+struct {
+    opaque server_id<1..2^16-1>;
+    uint16 ciphersuite;
+    opaque verification_key<1..2^16-1>;
+    KeyCommitment commitments<1..2^16-1>;
+}
+~~~
+
+5. The client then verifies the signature for each key commitment and
+stores the list of commitments to the current scope. The client SHOULD
+not cache the commitments beyond the current scope, as new commitments
+should be fetched for each independent issuance and redemption request.
+
+# Privacy Pass Issuance {#issuance}
+
+Inputs:
+- ``server_origin``: The origin to request token issuance from.
+- ``count``: The number of tokens to request issuance for.
+
+Outputs:
+- ``tokens``: A list of tokens that have been signed via the Privacy
+  Pass protocol.
+
+1. When a client wants to request tokens from a server, it should first
+   fetch a key commitment from the issuer via the process described in
+   {{key-commitment}} and keep the result as ``commitment``.
+
+2. The client should then call the ``CLIENT_ISSUE_GEN`` interface
+   requesting ``count`` tokens storing the resulting ``c_data``,
+   ``i_data``, ``g_data``.
+
+3. The client then makes a POST request to
+   <``server_origin``>/.well-known/privacy-pass with a message of type
+   ``request-issuance`` and a body of:
+
+~~~
+enum { Normal(0) } IssuanceType;
+
+struct {
+    IssuanceType type = 0;
+    opaque issue_data<0..2^16-1> = client_issue.issue_data;
+}
+~~~
+
+4. The server, upon receipt of the ``request`` should call the
+   ``SERVER_ISSUE`` interface with the value of ``issue_data`` with a
+   result of ``server_issue_resp``.
+
+
+5. The server should then respond to the POST request with a message of
+   type ``issue`` and a body of:
+
+~~~
+struct {
+    IssuanceType type = request.type;
+    uint16 ciphersuite = server_issue_resp.ciphersuite;
+    opaque evals<1..2^16-1> = server_issue_resp.evals;
+    opaque proof<1..2^16-1> = server_issue_resp.proof;
+}
+~~~
+
+6. The client should then should call the ``CLIENT_ISSUE_FINISH``
```

Comment:
This is slightly different to how the SERVER_ISSUE interface currently operates, as it this interface that calls CLIENT_ISSUE_FINISH. Perhaps the interface should be modified so that SERVER_ISSUE returns the response to CLIENT_ISSUE_GEN to match what you have here?

Reply:
Yeah, was going to comment on the architecture doc, in that for some of the interfaces its a bit odd that they internally call back and forth between the server and client, which makes it hard to integrate at the communication between the two.

One potential change is making them return these intermediate responses which are then used as inputs to the matching side, and optionally having interfaces that encapsulate the whole flow, so that APIs built on top of them can integrate between the individual steps rather than needing to redefine how they internally call back and forth.

Reply:
Yeah rewriting the interfaces to return those values synchronously to the calling interface makes sense. I'll write up the change.